File: em Java/Lucas_Victor_Nunes_AnalisadorSintatico/src/main/java/com/fesa/lucas_victor_nunes_analisadorsintatico/Analisador.java
package com.fesa.lucas_victor_nunes_analisadorsintatico;

import java.util.Stack;
import java.util.List;
import java.util.ArrayList;

public class Analisador {
    private Stack<Character> pilha = new Stack<>();
    private String input;
    private int indiceAtual;

    private List<Character> terminais = new ArrayList<>();
    private List<Character> variaveis = new ArrayList<>();

    private String[][] tabelaTransicoes = {
        { "Aa", "Aa", "Aa" },       // S
        { "", "BD", "BD" },         // A
        { "-",  "b", "" },          // B
        { "",  "-", "d" },           // D
    };

    public boolean realizaAnalise(String input) {
        iniciaAnalisador(input);
        return validaInput();
    }

    private void iniciaAnalisador(String input) {
        pilha.clear();
        pilha.push('S');

        indiceAtual = 0;
        this.input = input.toLowerCase();
        
        terminais.add('a');
        terminais.add('b');
        terminais.add('d');
        
        variaveis.add('S');
        variaveis.add('A');
        variaveis.add('B');
        variaveis.add('D');
    }

    private boolean validaInput() {
        char simboloAtual;

        int posVariavel;
        int posTerminal;
        String valorEmpilhar;

        while (!pilha.isEmpty()) {
            simboloAtual = pilha.pop();

            if (verificaSeContem(simboloAtual, terminais)) {
                if (indiceAtual >= input.length() || input.charAt(indiceAtual) != simboloAtual)
                    return false;

                indiceAtual++;
            } else if (verificaSeContem(simboloAtual, variaveis)) {
                posVariavel = retornaPosicaoLista(simboloAtual, variaveis);

                if (indiceAtual >= input.length())
                    posTerminal = tabelaTransicoes[0].length - 1;
                else
                    posTerminal = retornaPosicaoLista(input.charAt(indiceAtual), terminais);

                if (posTerminal == -1)
                    return false;

                valorEmpilhar = tabelaTransicoes[posVariavel][posTerminal];

                if (valorEmpilhar.equals("-"))
                    return false;

                for (int i = valorEmpilhar.length() - 1; i >= 0; i--) {
                    pilha.push(valorEmpilhar.charAt(i));
                }
            } else {
                return false;
            }
        }

        return indiceAtual == input.length();
    }

    private boolean verificaSeContem(char simbolo, List<Character> lista) {
        return lista.contains(simbolo);
    }

    private int retornaPosicaoLista(char simbolo, List<Character> lista) {
        return lista.indexOf(simbolo);
    }
}

File: in C#/AnalisadorSintatico_Lucas_Victor/Analisador.cs
namespace AnalisadorSintatico
{
    internal class Analisador
    {
        private Stack<char> pilha = new Stack<char>();
        private string input;
        private int indiceAtual;

        private List<char> terminais = new List<char>() { 'a', 'b', 'd' };
        private List<char> variaveis = new List<char>() { 'S', 'A', 'B', 'D' };

        private string[,] tabelaTransicoes =
        {
            { "Aa", "Aa", "Aa" },       // S
            { "", "BD", "BD" },         // A
            { "-",  "b", "" },          // B
            { "",  "-", "d" },          // D
        };

        public bool RealizaAnalise(string input)
        {
            IniciaAnalisador(input);
            return ValidaInput();
        }

        private void IniciaAnalisador(string input)
        {
            pilha.Clear();
            pilha.Push('S');

            indiceAtual = 0;
            this.input = input.ToLower();
        }

        private bool ValidaInput()
        {
            char simboloAtual;

            int posVariavel;
            int posTerminal;
            string valorEmpilhar;

            while(pilha.Count > 0)
            {
                simboloAtual = pilha.Pop();

                if (VerificaSeContem(simboloAtual, terminais))
                {
                    // Se o símbolo que estava no topo da pilha for terminal, ele obrigatoriamente deve ser igual do indice atual
                    if (indiceAtual >= input.Length || input[indiceAtual] != simboloAtual)
                        return false;

                    // Avanço o cursor
                    indiceAtual++;
                }
                else if (VerificaSeContem(simboloAtual, variaveis))
                {
                    // Válido quais simbolos serão empilhados no seu lugar a partir do input atual
                    posVariavel = RetornaPosicaoLista(simboloAtual, variaveis);

                    // Se eu já tiver terminado meu input, pego a posição do fim da pilha ($)
                    if (indiceAtual >= input.Length)
                        posTerminal = tabelaTransicoes.GetLength(1) - 1;
                    else
                        posTerminal = RetornaPosicaoLista(input[indiceAtual], terminais);

                    if (posTerminal == -1)
                        return false;

                    valorEmpilhar = tabelaTransicoes[posVariavel, posTerminal];

                    if(valorEmpilhar == "-")
                        return false;

                    for (int i = valorEmpilhar.Length - 1; i >= 0; i--)
                        pilha.Push(valorEmpilhar[i]);
                }
                else
                {
                    return false;
                }
            }

            return indiceAtual == input.Length;
        }

        private bool VerificaSeContem(char simbolo, List<char> lista)
        {
            return lista.Contains(simbolo);
        }

        private int RetornaPosicaoLista(char simbolo, List<char> lista)
        {
            return lista.IndexOf(simbolo);
        }
    }
}